src/MongoDB.Bson/IO/BsonBinaryReader.cs

```diff
@@ -121,9 +121,35 @@ public override bool IsAtEndOfFile()
         /// Reads BSON binary data from the reader.
         /// </summary>
         /// <returns>A BsonBinaryData.</returns>
-#pragma warning disable 618 // about obsolete BsonBinarySubType.OldBinary
         public override BsonBinaryData ReadBinaryData()
         {
+#pragma warning disable 618
+            var binaryData = ReadBinaryDataIgnoringGuidRepresentation();
+
+            if (BsonDefaults.GuidRepresentationMode == GuidRepresentationMode.V2)
+            {
+                var bytes = binaryData.Bytes;
+                var subType = binaryData.SubType;
+
+                GuidRepresentation guidRepresentation;
+                switch (subType)
+                {
+                    case BsonBinarySubType.UuidLegacy: guidRepresentation = _settings.GuidRepresentation; break;
+                    case BsonBinarySubType.UuidStandard: guidRepresentation = GuidRepresentation.Standard; break;
+                    default: guidRepresentation = GuidRepresentation.Unspecified; break;
+                }
+                binaryData = new BsonBinaryData(bytes, subType, guidRepresentation);
+            }
+
+            return binaryData;
+#pragma warning restore 618
+        }
+
+        /// <inheritdoc/>
+        [Obsolete("Use ReadBinaryData instead once you have transitioned to GuidRepresentationMode V3.")]
+        public override BsonBinaryData ReadBinaryDataIgnoringGuidRepresentation()
```

Comment:
This method is just like ReadBinaryData except that it ignores the GuidRepresentation setting.